### ISSUE TITLE
你好，请问一下我训练了200步，然后推理出的结果都是乱码，是不是训练步数不够？

### ISSUE THREAD

**Issue by guoyitao** [DCDmllm/Heartcare-Suite#3]

**Comment by guoyitao** [comment]
`import os
os.environ["CUDA_VISIBLE_DEVICES"] = "0"
os.environ["HF_DATASETS_OFFLINE"] = "1"
os.environ["HF_HUB_OFFLINE"] = "1"
# os.environ["PYTORCH_CUDA_ALLOC_CONF"] = "expandable_segments:True"

import pandas as pd
import numpy as np
import random
import math
import ast

import torch
from torch.utils.data import DataLoader, TensorDataset
import utils.my_ecg_process as ecg
from utils.my_tokenizer import Tokenizer
from utils.my_templates import Sentences, Choices, Reports, Predict

from transformers import AutoModelForCausalLM, AutoTokenizer, AutoConfig, TrainingArguments, set_seed
from safetensors.torch import load_file
from peft import LoraConfig, TaskType
from trl import SFTTrainer
from datasets import Dataset

def message_form(template):
    user = {
        "content": f"{template[0]}\n{template[1]}",
        "role": "user"
    }

    assistant = {
        "content": template[2],
        "role": "assistant"
    }

    msg = {"text": text_tokenizer.apply_chat_template([user, assistant], add_generation_prompt=False, tokenize=False)}
    return msg

symptoms = ["diagnostic", "form", "rhythm"]

def create_message_column(row):
    messages = []
    ecg_text = [f"<|ecg_{row.tokens[i][0]+1}|>" for i in range(len(row.tokens))]
    ecg_input = "<|ecg_start|> " + " ".join(ecg_text) + " <|ecg_end|>"
    age = int(row.age)
    sex = row.sex

    symptom_dict_list = [ast.literal_eval(row[symptoms[idx]]) for idx in range(3)]
    S = Sentences.get_template(age, sex, ecg_input, symptom_dict_list)
    C = Choices.get_template(age, sex, ecg_input, symptom_dict_list)
    R = Reports.get_template(age, sex, ecg_input, symptom_dict_list)
    P = Predict.get_template(age, sex, ecg_input)
    lists = S + C + R + P
    messages = [message_form(l) for l in lists]
    return messages


# set_seed(42)
print(f"CUDA_VISIBLE_DEVICES: {os.environ.get('CUDA_VISIBLE_DEVICES')}")
print(f"PyTorch sees {torch.cuda.device_count()} GPU(s)")
print(f"Current device index: {torch.cuda.current_device()}")
print(f"Device name: {torch.cuda.get_device_name(torch.cuda.current_device())}")


sampling_rate = 500
batch_size = 32
seq_length = 500
patch_size = 25
latent_ratio = 0.5
channels = 12
codebook_size = 256
residual_levels = 2
device = torch.device('cuda:0' if torch.cuda.is_available() else 'cpu')
dir = f"level_{residual_levels}_code_{codebook_size}_len_{seq_length}_ratio_{latent_ratio}"
ecg_tokenizer_path = f"tokenizer/ecg_tokenizer_{dir}.pth"
model_path = 'C:\mycode\yiliao\Phi-3-mini-4k-instruct'
num = 10
trainer_path = f"training/{num}"
merged_model_path = f"phi-3-ecg/{num}"


vq_kwargs = {'residual_levels': residual_levels}
ecg_tokenizer = Tokenizer(
    seq_length=seq_length,
    patch_size=patch_size,
    latent_ratio=latent_ratio,
    channels=channels,
    codebook_size=codebook_size,
    vq_kwargs=vq_kwargs
).to(device)
# ecg_weights = torch.load(ecg_tokenizer_path, weights_only=False)
# ecg_tokenizer.load_state_dict(ecg_weights['model_state_dict'])


tok = {
    'model_state_dict': ecg_tokenizer.state_dict(),
}
torch.save(tok, f"tokenizer/tok_2.pth")

config = AutoConfig.from_pretrained(
    model_path,
    trust_remote_code=False
)

text_tokenizer = AutoTokenizer.from_pretrained(
    model_path,
    trust_remote_code=False
)

text_tokenizer.pad_token = text_tokenizer.unk_token
text_tokenizer.pad_token_id = text_tokenizer.convert_tokens_to_ids(text_tokenizer.pad_token)
text_tokenizer.padding_side = 'left'


model = AutoModelForCausalLM.from_pretrained(
    model_path,
    config=config,
    trust_remote_code=False,
    torch_dtype=torch.float16,
    attn_implementation='sdpa',
    # device_map="auto"
).to(device)

weight1 = load_file(f"{model_path}/model-00001-of-00002.safetensors")
weight2 = load_file(f"{model_path}/model-00002-of-00002.safetensors")

state_dict = {**weight1, **weight2}
model.load_state_dict(state_dict, strict=False)




peft_config = LoraConfig(
    r=64,
    lora_alpha=64,
    lora_dropout=0.05,
    task_type=TaskType.CAUSAL_LM,
    target_modules=['k_proj', 'q_proj', 'v_proj', 'o_proj', "gate_proj", "down_proj", "up_proj"],
)

args = TrainingArguments(
    output_dir=trainer_path,
    logging_dir=f"runs/{trainer_path}",
    save_strategy="steps",
    save_steps=50,
    save_total_limit=1,
    eval_strategy="steps",
    eval_steps=50,
    do_eval=True,
    optim="adamw_torch",
    per_device_train_batch_size=8,
    per_device_eval_batch_size=8,
    log_level="warning",
    logging_steps=10,
    learning_rate=1e-4,
    fp16=False,
    bf16=False,
    num_train_epochs=3,
    warmup_ratio=0.0,
    lr_scheduler_type="linear",
    report_to="tensorboard",
    # seed=42,
    max_steps=200,  # 最多训练 100 步
)


if __name__ == '__main__':
    X = np.load("data/records500.npy")
    total_length = int(seq_length * (1 + latent_ratio))
    start = int((X.shape[2] - total_length)/2)
    signal = X[:, :, start:start+seq_length]
    predict = X[:, :, start+seq_length:start+total_length]

    df_dataset = TensorDataset(torch.tensor(signal, dtype=torch.float32))
    df_loader = DataLoader(df_dataset, batch_size=batch_size, shuffle=True, num_workers=4, pin_memory=True)


    tokens = []
    new = []
    with torch.no_grad():
        for batch in df_loader:
            x = batch[0].to(device)
            ecg_tokens = ecg_tokenizer.tokenize(x).cpu().detach().tolist()
            tokens.extend(ecg_tokens)


    combined_df = pd.read_parquet("data/full_data.parquet")


    X = np.load("data/records500.npy")
    total_length = int(seq_length * (1 + latent_ratio))
    start = int((X.shape[2] - total_length)/2)
    signal = X[:, :, start:start+seq_length]
    predict = X[:, :, start+seq_length:start+total_length]

    df_dataset = TensorDataset(torch.tensor(signal, dtype=torch.float32))
    df_loader = DataLoader(df_dataset, batch_size=batch_size, shuffle=True, num_workers=4, pin_memory=True)

#    tokens = []
#    new = []
#    with torch.no_grad():
#        for batch in df_loader:
#            x = batch[0].to(device)
#            ecg_tokens = ecg_tokenizer.tokenize(x).cpu().detach().tolist()
#            tokens.extend(ecg_tokens)

#    combined_df["tokens"] = tokens


    new_tokens = ["<|ecg_start|>", "<|ecg_end|>", "<|report_start|>", "<|report_end|>", "<|pred_start|>", "<|pred_end|>"] + [f"<|ecg_{i+1}|>" for i in range(codebook_size)]
    text_tokenizer.add_tokens(new_tokens)
    model.resize_token_embeddings(len(text_tokenizer))

    dataset_chatml = combined_df.apply(create_message_column, axis=1)


    valid_fold = 9
    test_fold = 10
    train_list = dataset_chatml[(combined_df.strat_fold != valid_fold) & (combined_df.strat_fold != test_fold)]
    valid_list = dataset_chatml[combined_df.strat_fold == valid_fold]
    test_list = dataset_chatml[combined_df.strat_fold == test_fold]

    # .shuffle(seed=42)
    train_dataset = Dataset.from_list([item for sublist in train_list for item in sublist])
    valid_dataset = Dataset.from_list([item for sublist in valid_list for item in sublist])

    # test_dataset = Dataset.from_list([item for sublist in test_list for item in sublist])

    # 取前 1000 条训练数据
    train_dataset_small = train_dataset.select(range(1000))

    # 取前 200 条验证数据
    valid_dataset_small = valid_dataset.select(range(200))


    trainer = SFTTrainer(
        model=model,
        train_dataset=train_dataset_small,
        eval_dataset=valid_dataset_small,
        peft_config=peft_config,
        tokenizer=text_tokenizer,
        dataset_text_field="text",
        max_seq_length=512,
        args=args,
    )

    trainer.train()
    model.resize_token_embeddings(len(text_tokenizer))
    model.save_pretrained(merged_model_path)
    text_tokenizer.save_pretrained(merged_model_path)

`

**Comment by guoyitao** [comment]
```
import os
os.environ["CUDA_VISIBLE_DEVICES"] = "0"
os.environ["HF_DATASETS_OFFLINE"] = "1"
os.environ["HF_HUB_OFFLINE"] = "1"

import pandas as pd
import numpy as np
import random
import math
import ast

import torch
from torch.utils.data import DataLoader, TensorDataset
import utils.my_ecg_process as ecg
from utils.my_tokenizer import Tokenizer
from utils.my_templates import Sentences, Choices, Reports, Predict

from transformers import AutoModelForCausalLM, AutoTokenizer, AutoConfig, TrainingArguments, set_seed
from safetensors.torch import load_file
from peft import LoraConfig, TaskType
from trl import SFTTrainer
from datasets import Dataset

def message_form(template):
    user = {
        "content": f"{template[0]}\n{template[1]}",
        "role": "user"
    }

    assistant = {
        "content": template[2],
        "role": "assistant"
    }

    msg = {"text": text_tokenizer.apply_chat_template([user, assistant], add_generation_prompt=False, tokenize=False)}
    return msg

symptoms = ["diagnostic", "form", "rhythm"]

def create_message_column(row):
    messages = []
    ecg_text = [f"<|ecg_{row.tokens[i][0]+1}|>" for i in range(len(row.tokens))]
    ecg_input = "<|ecg_start|> " + " ".join(ecg_text) + " <|ecg_end|>"
    age = int(row.age)
    sex = row.sex

    symptom_dict_list = [ast.literal_eval(row[symptoms[idx]]) for idx in range(3)]
    S = Sentences.get_template(age, sex, ecg_input, symptom_dict_list)
    C = Choices.get_template(age, sex, ecg_input, symptom_dict_list)
    R = Reports.get_template(age, sex, ecg_input, symptom_dict_list)
    P = Predict.get_template(age, sex, ecg_input)
    lists = S + C + R + P
    messages = [message_form(l) for l in lists]
    return messages


print(f"CUDA_VISIBLE_DEVICES: {os.environ.get('CUDA_VISIBLE_DEVICES')}")
print(f"PyTorch sees {torch.cuda.device_count()} GPU(s)")
print(f"Current device index: {torch.cuda.current_device()}")
print(f"Device name: {torch.cuda.get_device_name(torch.cuda.current_device())}")


sampling_rate = 500
batch_size = 32
seq_length = 500
patch_size = 25
latent_ratio = 0.5
channels = 12
codebook_size = 256
residual_levels = 2
device = torch.device('cuda:0' if torch.cuda.is_available() else 'cpu')
dir = f"level_{residual_levels}_code_{codebook_size}_len_{seq_length}_ratio_{latent_ratio}"
ecg_tokenizer_path = f"tokenizer/ecg_tokenizer_{dir}.pth"
model_path = 'C:\mycode\yiliao\Phi-3-mini-4k-instruct'
num = 10
trainer_path = f"training/{num}"
merged_model_path = f"phi-3-ecg/{num}"


vq_kwargs = {'residual_levels': residual_levels}
ecg_tokenizer = Tokenizer(
    seq_length=seq_length,
    patch_size=patch_size,
    latent_ratio=latent_ratio,
    channels=channels,
    codebook_size=codebook_size,
    vq_kwargs=vq_kwargs
).to(device)

tok = {
    'model_state_dict': ecg_tokenizer.state_dict(),
}
torch.save(tok, f"tokenizer/tok_2.pth")

config = AutoConfig.from_pretrained(
    model_path,
    trust_remote_code=False
)

text_tokenizer = AutoTokenizer.from_pretrained(
    model_path,
    trust_remote_code=False
)

text_tokenizer.pad_token = text_tokenizer.unk_token
text_tokenizer.pad_token_id = text_tokenizer.convert_tokens_to_ids(text_tokenizer.pad_token)
text_tokenizer.padding_side = 'left'


model = AutoModelForCausalLM.from_pretrained(
    model_path,
    config=config,
    trust_remote_code=False,
    torch_dtype=torch.float16,
    attn_implementation='sdpa',
    # device_map="auto"
).to(device)

weight1 = load_file(f"{model_path}/model-00001-of-00002.safetensors")
weight2 = load_file(f"{model_path}/model-00002-of-00002.safetensors")

state_dict = {**weight1, **weight2}
model.load_state_dict(state_dict, strict=False)




peft_config = LoraConfig(
    r=64,
    lora_alpha=64,
    lora_dropout=0.05,
    task_type=TaskType.CAUSAL_LM,
    target_modules=['k_proj', 'q_proj', 'v_proj', 'o_proj', "gate_proj", "down_proj", "up_proj"],
)

args = TrainingArguments(
    output_dir=trainer_path,
    logging_dir=f"runs/{trainer_path}",
    save_strategy="steps",
    save_steps=50,
    save_total_limit=1,
    eval_strategy="steps",
    eval_steps=50,
    do_eval=True,
    optim="adamw_torch",
    per_device_train_batch_size=8,
    per_device_eval_batch_size=8,
    log_level="warning",
    logging_steps=10,
    learning_rate=1e-4,
    fp16=False,
    bf16=False,
    num_train_epochs=3,
    warmup_ratio=0.0,
    lr_scheduler_type="linear",
    report_to="tensorboard",
    max_steps=200,  # 最多训练 100 步
)


if __name__ == '__main__':
    X = np.load("data/records500.npy")
    total_length = int(seq_length * (1 + latent_ratio))
    start = int((X.shape[2] - total_length)/2)
    signal = X[:, :, start:start+seq_length]
    predict = X[:, :, start+seq_length:start+total_length]

    df_dataset = TensorDataset(torch.tensor(signal, dtype=torch.float32))
    df_loader = DataLoader(df_dataset, batch_size=batch_size, shuffle=True, num_workers=4, pin_memory=True)


    tokens = []
    new = []
    with torch.no_grad():
        for batch in df_loader:
            x = batch[0].to(device)
            ecg_tokens = ecg_tokenizer.tokenize(x).cpu().detach().tolist()
            tokens.extend(ecg_tokens)


    combined_df = pd.read_parquet("data/full_data.parquet")


    X = np.load("data/records500.npy")
    total_length = int(seq_length * (1 + latent_ratio))
    start = int((X.shape[2] - total_length)/2)
    signal = X[:, :, start:start+seq_length]
    predict = X[:, :, start+seq_length:start+total_length]

    df_dataset = TensorDataset(torch.tensor(signal, dtype=torch.float32))
    df_loader = DataLoader(df_dataset, batch_size=batch_size, shuffle=True, num_workers=4, pin_memory=True)



    new_tokens = ["<|ecg_start|>", "<|ecg_end|>", "<|report_start|>", "<|report_end|>", "<|pred_start|>", "<|pred_end|>"] + [f"<|ecg_{i+1}|>" for i in range(codebook_size)]
    text_tokenizer.add_tokens(new_tokens)
    model.resize_token_embeddings(len(text_tokenizer))

    dataset_chatml = combined_df.apply(create_message_column, axis=1)


    valid_fold = 9
    test_fold = 10
    train_list = dataset_chatml[(combined_df.strat_fold != valid_fold) & (combined_df.strat_fold != test_fold)]
    valid_list = dataset_chatml[combined_df.strat_fold == valid_fold]
    test_list = dataset_chatml[combined_df.strat_fold == test_fold]


    train_dataset = Dataset.from_list([item for sublist in train_list for item in sublist])
    valid_dataset = Dataset.from_list([item for sublist in valid_list for item in sublist])


    # 取前 1000 条训练数据
    train_dataset_small = train_dataset.select(range(1000))

    # 取前 200 条验证数据
    valid_dataset_small = valid_dataset.select(range(200))


    trainer = SFTTrainer(
        model=model,
        train_dataset=train_dataset_small,
        eval_dataset=valid_dataset_small,
        peft_config=peft_config,
        tokenizer=text_tokenizer,
        dataset_text_field="text",
        max_seq_length=512,
        args=args,
    )

    trainer.train()
    model.resize_token_embeddings(len(text_tokenizer))
    model.save_pretrained(merged_model_path)
    text_tokenizer.save_pretrained(merged_model_path)

> 
```

**Comment by LoyaltyLo** [comment]
你好，我想问一下你这边是怎么去使用的，我感觉他这边代码不太完整

**Comment by wwj-2017-1117** [comment]
一起探索下？

**Comment by zyfjrx** [comment]
这个目前有源码吗？

**Comment by SHUSHENGQIGUI** [comment]
where is data/records250/records250.npy?

**Comment by SHUSHENGQIGUI** [comment]
> 一起探索下？

data/records250/records250.npy 这个数据集从哪获取呢
